### ISSUE TITLE
add pause and resume

### ISSUE THREAD

**Issue by quanvo87** [dingwilson/SwiftVideoBackground#54]
Scenario:
- a video is being played on a `UIView`
- that `UIView` becomes hidden (ie another view controller is pushed onto the stack)
- that `UIView` becomes un-hidden again (ie previously mentioned view controller is closed)

What happens:
It seems that while `UIView` was hidden, the video wasn't playing, and when it gets un-hidden, it plays the video super fast, up to the point that it would be if it wasn't paused.

Not sure if that makes sense or if that's actually what's happening, but I know there is some super fast video playing after a hidden/un-hidden scenario.

I think a fix would be to implement `pause()` and `resume()`. I'm hoping correctly using these methods when you expect your view to be hidden/un-hidden fixes this.

This might also be able to be automated if desired (not sure if possible, not sure of side effects).

**Comment by quanvo87** [comment]
After some research, automating this may be a bad idea. There are too many scenarios where a `UIView` playing a video may/may not want to be paused:
- another view is pushed on top of it
- another view partially covers it, at what threshold do you pause?
- another view is pushed on top, at what transparency of this other view threshold do you pause?
- the view is in a cell which is off screen

tldr I'll leave automation out.

**Comment by dingwilson** [comment]
Your points make sense. Lets still implement `pause()` and `resume()`, as I can see definite use cases for those. It shouldn't be too difficult.

**Comment by quanvo87** [comment]
Yes, that way users can manually pause and resume when they want, ie when their view will be offscreen/covered/whatever the case.

I'm going to do this after I add support for playing videos on multiple views with a single instance of `SVB`.

The idea is:
- anytime a user wants to play a vid on a view, they basically call:
  - `VideoBackground.shared.play(view, video)` (this is already implemented)
- to play another vid on another view, they would be able to call: 
  - `VideoBackground.shared.play(anotherView, anotherVideo)` (I'm going to implement this)
- to pause the first video, they would have to call `pause(view)`
- to resume: `resume(view)`

You have to pass in a view to specify on which view to pause/resume the vid. If you pass in a view that hasn't been previously passed into `SVB`, nothing happens.

Under the covers, whenever a user calls `play(view, video`), we will add that view and a new `metadata struct/class` that will keep track of info for that view, such as `isMute`, the observers, etc. The view will be a key, and metadata will be a value, in a [NSMapTable](https://developer.apple.com/documentation/foundation/nsmaptable), which is basically a dictionary with memory retainment options for its key/values (with auto clean up when a key/value is released).

This is my initial implementation idea. Let me know of comments/criticisms.

None of these are breaking changes and can be released as a minor patch.

**Comment by quanvo87** [comment]
Regarding the singleton aspect, after multiview support is added, there is no reason to instantiate your own instance of `SVB` (and it may even be able to break something if you do and start doing weird things with it). Thus we can deprecate `init`, and in the next major release make it private.

**Comment by dingwilson** [comment]
That looks great! Good call on keeping track of the metadata... I didn't think of that haha.

**Comment by quanvo87** [comment]
I'm also thinking (I guess I'm bored af/a nerd), that we can even move the APIs into an extension of `UIView`, so that you could just call `view.play(video)`.

The initial problem with this is that where do we retain the observers required for this view? If we could access a singleton retaining this `NSMapTable` in these extensions, then that should be no problem, and the API is even cleaner.

This would not break anything either, just provide two ways of doing the same thing, with one API very clean. This is a bit confusing, and if worth implementing, should be done cautiously and extra mindfully.

**Comment by dingwilson** [comment]
That would be beautiful.

**Comment by quanvo87** [comment]
Interestingly I am not seeing the speed up on an actual device. Maybe just a simulator issue. Nonetheless I'll add pause() and resume().
